### PR TITLE
[Merged by Bors] - feat: port CategoryTheory.Monoidal.Free.Coherence

### PR DESCRIPTION
**Pull request by joelriou**
---
- [x] depends on: #2808

See https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/!4.232808.20CategoryTheory.2EMonoidal.2EFree.2EBasic/near/355402986

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#2808
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/CategoryTheory/Monoidal/Free/Coherence.lean [hunk 1]
```diff
@@ -56,28 +56,28 @@ variable (C)
...
--- mathport name: exprF
 local notation "F" => FreeMonoidalCategory
 
--- mathport name: exprN
 local notation "N" => Discrete ∘ NormalMonoidalObject
 
--- mathport name: «expr ⟶ᵐ »
 local infixr:10 " ⟶ᵐ " => Hom
 
+-- porting note: this was automatic in mathlib
```
Comment (jcommelin) [inline Mathlib/CategoryTheory/Monoidal/Free/Coherence.lean:73]:
```suggestion
-- porting note: this was automatic in mathlib 3
```

**Review by jcommelin — COMMENTED** [review]
bors d+

File: Mathlib/CategoryTheory/Monoidal/Free/Coherence.lean [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+#align category_theory.free_monoidal_category.discrete_functor_obj_eq_as CategoryTheory.FreeMonoidalCategory.discrete_functor_obj_eq_as
+
+-- TODO: move to discrete_category.lean, decide whether this should be a global simp lemma
+@[simp 1100]
+theorem discrete_functor_map_eq_id (g : X ⟶ X) : (Discrete.functor f).map g = 𝟙 _ := rfl
+#align category_theory.free_monoidal_category.discrete_functor_map_eq_id CategoryTheory.FreeMonoidalCategory.discrete_functor_map_eq_id
+
+end
+
+set_option maxHeartbeats 400000 in
```
Comment (mattrobball) [inline Mathlib/CategoryTheory/Monoidal/Free/Coherence.lean:242]: Is this still necessary with #3414?

File: Mathlib/CategoryTheory/Monoidal/Free/Coherence.lean [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+section
+
+open Hom
+
+--attribute [local tidy] tactic.discrete_cases
+
+-- porting note: triggers a PANIC "invalid LCNF substitution of free variable
+-- with expression CategoryTheory.FreeMonoidalCategory.NormalMonoidalObject.{u}"
+-- prevented with dsimp [normalizeObj]
+-- the @[simp] attribute is removed because it also triggers a PANIC
```
Comment (mattrobball) [inline Mathlib/CategoryTheory/Monoidal/Free/Coherence.lean:117]: Can this be restored now?

File: Mathlib/CategoryTheory/Monoidal/Free/Coherence.lean [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+
+section
+
+open Hom
+
+--attribute [local tidy] tactic.discrete_cases
+
+-- porting note: triggers a PANIC "invalid LCNF substitution of free variable
+-- with expression CategoryTheory.FreeMonoidalCategory.NormalMonoidalObject.{u}"
+-- prevented with dsimp [normalizeObj]
```
Comment (mattrobball) [inline Mathlib/CategoryTheory/Monoidal/Free/Coherence.lean:116]: Is `dsimp` enough to prevent the panics?

**Comment by mattrobball** [comment]
Yes `dsimp` does suffice. No you cannot restore the `@[simp]` :(

**Comment by mattrobball** [comment]
And the bump to max heartbeats can indeed be removed after #3414

**Pull request closed without merging**